src/render/renderer.cpp

```diff
@@ -167,31 +167,44 @@ glm::vec4 Renderer::traceRayMIP(const Ray& ray, float sampleStep) const
     return glm::vec4(glm::vec3(maxVal) / m_pVolume->maximum(), 1.0f);
 }
 
-// ======= TODO: IMPLEMENT ========
+// ======= TODO: check phong parameters ========
 // This function should find the position where the ray intersects with the volume's isosurface.
 // If volume shading is DISABLED then simply return the isoColor.
 // If volume shading is ENABLED then return the phong-shaded color at that location using the local gradient (from m_pGradientVolume).
 //   Use the camera position (m_pCamera->position()) as the light position.
 // Use the bisectionAccuracy function (to be implemented) to get a more precise isosurface location between two steps.
 glm::vec4 Renderer::traceRayISO(const Ray& ray, float sampleStep) const
 {
-    glm::vec3 sample_pos = ray.origin + ray.tmin * ray.direction;
-    const glm::vec3 increment = sampleStep * ray.direction;
     glm::vec4 color(0.0f);
+    if (!this->m_config.volumeShading) {
+        glm::vec3 sample_pos = ray.origin + ray.tmin * ray.direction;
+        const glm::vec3 increment = sampleStep * ray.direction;
+
+        for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, sample_pos += increment) {
+            auto voxel_value = this->m_pVolume->getVoxelInterpolate(sample_pos);
+            if (voxel_value > this->m_config.isoValue) {
+                color = this->getTFValue(voxel_value);
+                // color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
+                break;
+            }
+        }
+    } else {
+
+        glm::vec3 sample_pos = ray.origin + ray.tmin * ray.direction;
+        const glm::vec3 increment = sampleStep * ray.direction;
 
-    for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, sample_pos += increment) {
-        auto voxel_value = this->m_pVolume->getVoxelInterpolate(sample_pos);
-        if (voxel_value > this->m_config.isoValue) {
-            color = this->getTFValue(voxel_value);
-            // color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
-            break;
+        for (float t = ray.tmin; t <= ray.tmax; t += sampleStep, sample_pos += increment) {
+            auto voxel_value = this->m_pVolume->getVoxelInterpolate(sample_pos);
+            if (voxel_value > this->m_config.isoValue) {
+                auto _color = glm::vec4 { 0.8f, 0.8f, 0.2f, 1.0f };
+                auto gradient = this->m_pGradientVolume->getGradientVoxel(sample_pos);
+                color = glm::vec4(this->computePhongShading(_color, gradient, this->m_pCamera->position(), this->m_pCamera->position()), 1.0f);

```

Comment:
why is the 'view' vector in the signature of phongshading the same as the light vector?

Reply:
I still left a TODO regarding this. I'm still not 100% sure how to get that parameter right.